engines/dfc_provider/app/services/dfc_provider/authorization_control.rb

```diff
@@ -8,24 +8,51 @@ def initialize(access_token)
       @access_token = access_token
     end
 
+    def safe_process
+      process
+    rescue JWT::VerificationError, JWT::ExpiredSignature, JWT::ImmatureSignature => e
+      # To help development debugging
+      raise e if Rails.env.development?
+
+      nil
+    end
+
     def process
       decode_token
+      control_payload
       find_ofn_user
     end
 
+    private
+
     def decode_token
       data = JWT.decode(
         @access_token,
         nil,
-        false
+        true,
+        { algorithms: ['RS256'], jwks: jwks_hash }
       )
 
       @header = data.last
       @payload = data.first
     end
 
     def find_ofn_user
-      Spree::User.where(email: @payload['email']).first
+      Spree::User.where(provider: 'openid_connect', uid: @payload['email']).first
+    end
+
+    def control_payload
+      raise 'Email Not Found' if @payload['email'].blank?
+    end
+
+    def jwks_uri
+      Devise.omniauth_configs[:openid_connect].options[:client_options][:jwks_uri]
+    end
+
+    def jwks_hash
+      jwks_raw = Net::HTTP.get URI(jwks_uri)
+
+      JSON.parse(jwks_raw)
```

Comment:
@Matt-Yorkley I was thinking to cache this into Memcached for sometime (1 hour), what do you think about it? We don't want to request the OIDC configuration everytime.

Reply:
Well for starters Memcached doesn't exist any more :stuck_out_tongue: 

I don't think caching JWKS tokens for an arbitrary amount of time is a good idea. But I agree it shouldn't be calling the OIDC for every request to the API :+1: